### ISSUE TITLE
failing save for versions 0.5.8 and 0.6.0 when using elasticsearch for persistance

### ISSUE THREAD

**Issue by gfrivolt** [karmi/retire#768]
I'm using elasticsearch/tire for persistance in the way as it is described in the documentation. For tire 0.5.7 I was able to save the models, however, for versions 0.5.8 and 0.6.0 I get this error:

```
NoMethodError: undefined method `[]' for #<Tire::Model::Search::InstanceMethodsProxy:0x007fdd07222e00>
```

The save seems to fail on the `!! response['ok']` line in `storage.rb`:

```
def save
  return false unless valid?
  run_callbacks :save do
    response = update_index
    !! response['ok']
  end
end
```

**Comment by karmi** [comment]
This has been introduced in 46fbcbf. Do the integration tests for persistence run for you? Can you post more info/backtrace?

**Comment by gfrivolt** [comment]
I executed the tests and found 4 failing for the current state of the repository. The output is [here](http://gist.github.com/fifigyuri/5858316).

I'm using this declaration for the model:

```
class Article
  include Tire::Model::Search
  include Tire::Model::Callbacks
  include Tire::Model::Persistence

  alias_attribute :content, :body

  property :author,       type: 'string'
  property :title,        type: 'string'
  property :summary,      type: 'string'
  property :body,         type: 'string'
  property :url,          type: 'string'
  property :published_at, type: 'date'
  property :full_page,    type: 'string'
end
```

The articles itself are created by a factory, which adds `id` to them:

```
require 'securerandom'

class ArticleFactory

  def self.build &block
    article = Article.new
    article.id = SecureRandom.urlsafe_base64
    article.instance_eval(&block) if block_given?
    article
  end

  def self.create &block
    article = build(&block)
    article.save
    article
  end
end
```

Saving model constructed this way passed well earlier.

**Comment by karmi** [comment]
> I executed the tests and found 4 failing for the current state of the repository. The output is here.

I have meant only the integration tests for the persistence :) None of them is failing.

> Saving model constructed this way passed well earlier.

Could you attach or pastie somewhere a full backtrace please?

**Comment by gfrivolt** [comment]
hmm.. the issue shows up for me only in the specs, saving in running rails works..

here is the pastie for the traceback from my failing specs: http://pastie.org/8078721

**Comment by karmi** [comment]
Aren't you mocking the response in an incompatible way, perhaps?

**Comment by gfrivolt** [comment]
no stubs/no mocks in the failing specs

**Comment by gfrivolt** [comment]
I looked at the code again, actually the issue appears anytime when I try to save the tire persisted model. I created a simple rails 4 app, on saving it throws the error. The test rails is [here](https://github.com/fifigyuri/tire_test).

```
  → rails c
  Loading development environment (Rails 4.0.0)
  irb(main):003:0> a = Article.new
  => #<Article:0x007fcb3a53d290>
  irb(main):004:0> a.title = 'title'
  irb(main):009:0> a.valid?
  => true
  irb(main):010:0> a.save
  NoMethodError: undefined method `[]' for #<Tire::Model::Search::InstanceMethodsProxy:0x007fcb3a6bf488>
          from /usr/local/opt/rbenv/versions/1.9.3-p392/lib/ruby/gems/1.9.1/gems/tire-0.6.0/lib/tire/model/persistence/storage.rb:60:in `block in save'
          from /usr/local/opt/rbenv/versions/1.9.3-p392/lib/ruby/gems/1.9.1/gems/activesupport-4.0.0/lib/active_support/callbacks.rb:373:in `_run__669920208294084740__save__callbacks'
          from /usr/local/opt/rbenv/versions/1.9.3-p392/lib/ruby/gems/1.9.1/gems/activesupport-4.0.0/lib/active_support/callbacks.rb:80:in `run_callbacks'
          from /usr/local/opt/rbenv/versions/1.9.3-p392/lib/ruby/gems/1.9.1/gems/tire-0.6.0/lib/tire/model/persistence/storage.rb:58:in `save'
          from (irb):10
          from /usr/local/opt/rbenv/versions/1.9.3-p392/lib/ruby/gems/1.9.1/gems/railties-4.0.0/lib/rails/commands/console.rb:90:in `start'
          from /usr/local/opt/rbenv/versions/1.9.3-p392/lib/ruby/gems/1.9.1/gems/railties-4.0.0/lib/rails/commands/console.rb:9:in `start'
          from /usr/local/opt/rbenv/versions/1.9.3-p392/lib/ruby/gems/1.9.1/gems/railties-4.0.0/lib/rails/commands.rb:64:in `<top (required)>'
          from bin/rails:4:in `require'
          from bin/rails:4:in `<main>'
```

**Comment by gfrivolt** [comment]
I included more modules, than necessary. Now I don't have any issues, as the correct callbacks are called. As this conflict of modules appeared only versions 0.5.8 and 0.6.0, I did not suspect my codes too much.

I included `Tire::Model::Search` and `Tire::Model::Callbacks` before `Tire::Model::Persistence`. Including `Tire::Model::Persistence` includes `Tire::Model::Search` as well. Sorry for the confusion, now at least I had to take a bit deeper look into the _tire_ codes.

A possible enhancement could be raising an error if `Tire::Model::Search` is included before `Tire::Model::Persistence`, otherwise I think this issue was a false alarm.
